p7zip: fix build on gcc 11

tobbez opened the pull request:
Fixes the following build failure when building with gcc 11:

         478    ../../../../CPP/7zip/Archive/Wim/WimHandler.cpp: In member function 'virtual LONG NArchive::NWim::CHandler::GetArchiveProperty(PROPID, PROPVARIANT*)':
      >> 479    ../../../../CPP/7zip/Archive/Wim/WimHandler.cpp:308:11: error: use of an operand of type 'bool' in 'operator++' is forbidden in C++17
         480      308 |           numMethods++;
         481          |           ^~~~~~~~~~
      >> 482    ../../../../CPP/7zip/Archive/Wim/WimHandler.cpp:318:9: error: use of an operand of type 'bool' in 'operator++' is forbidden in C++17
         483      318 |         numMethods++;
         484          |         ^~~~~~~~~~

vmiheer commented:
Sounds good. But we should be adding support for newer version (of 7zip) which would remove the requirement for such patches.
Note we wanted to update earlier but downloading source code of new version of 7zip needed 7zip and so it needed kind-of infrastructure changes.
The latest 7zip release seem to have .xz file, worth checking.

tobbez commented:
Is there anything I need to do here? The CI failure looks completely unrelated to the changes in this PR.

Long term I agree that the correct solution is to package (a modern version of) 7zip instead of (the no longer maintained) p7zip, but that essentially means packaging a new software (while p7zip is based on 7zip, they are not the same).

vmiheer commented:
I don't think anything is needed from you. I guess someone with merging privilege will eventually merge this :).
Thank you for your contribution 🙌

alalazo commented on var/spack/repos/builtin/packages/p7zip/package.py:
```diff
@@ -14,6 +14,7 @@ class P7zip(MakefilePackage):
     version('16.02', sha256='5eb20ac0e2944f6cb9c2d51dd6c4518941c185347d4089ea89087ffdd6e2341f')
 
     patch('gcc10.patch', when='%gcc@10:', sha256='96914025b9f431fdd75ae69768162d57751413634622f9df1a4bc4960e7e8fe1')
+    patch('gcc11.patch', when='%gcc@11:', sha256='39dd15f2dfc86eeee8c3a13ffde65c2ca919433cfe97ea126fbdc016afc587d1')
```
On a second thought, can you please add a brief comment above this directive to explain what the patch does and link to its provenance or a bug report if it exists?

tobbez replied:
Added a comment and rebased on top of develop.

There is no more provenance than is present in the commit. I discovered the issue when p7zip failed to build under GCC 11, and wrote the patch to fix it.

There is also no bug report, since the upstream project is no longer maintained (last release was in 2016).

tobbez committed:
p7zip: fix build on gcc 11

Fixes the following build failure when building with gcc 11:

         478    ../../../../CPP/7zip/Archive/Wim/WimHandler.cpp: In member function 'virtual LONG NArchive::NWim::CHandler::GetArchiveProperty(PROPID, PROPVARIANT*)':
      >> 479    ../../../../CPP/7zip/Archive/Wim/WimHandler.cpp:308:11: error: use of an operand of type 'bool' in 'operator++' is forbidden in C++17
         480      308 |           numMethods++;
         481          |           ^~~~~~~~~~
      >> 482    ../../../../CPP/7zip/Archive/Wim/WimHandler.cpp:318:9: error: use of an operand of type 'bool' in 'operator++' is forbidden in C++17
         483      318 |         numMethods++;
         484          |         ^~~~~~~~~~
